conbench/api/compare.py

```diff
@@ -160,53 +289,131 @@ def get(self, compare_ids) -> f.Response:
+        try:
+            comparator = BenchmarkResultComparator(
+                baseline=baseline_result,
+                contender=contender_result,
+                threshold=threshold,
+                threshold_z=threshold_z,
+            )
+        except UnmatchingUnitsError as e:
+            f.abort(400, description=str(e))
 
-        if baseline_benchmark_result.run.commit is None:
-            # TODO: emit a useful HTTP response
-            raise Exception("baseline commit unknown")
+        return f.jsonify(comparator._dict_for_api_json)
 
-        set_z_scores(
-            contender_benchmark_results=[contender_benchmark_result],
-            baseline_commit=baseline_benchmark_result.run.commit,
-        )
 
-        # TODO: remove baseline-z-score-related keys from this endpoint
-        baseline_benchmark_result.z_score = None
-
-        set_display_case_permutation(baseline_benchmark_result)
-        set_display_case_permutation(contender_benchmark_result)
-        set_display_benchmark_name(baseline_benchmark_result)
-        set_display_benchmark_name(contender_benchmark_result)
-
-        baseline = _result_dict_for_compare_api(baseline_benchmark_result)
-        contender = _result_dict_for_compare_api(contender_benchmark_result)
-        comparator = BenchmarkResultComparator(
-            baseline,
-            contender,
-            threshold,
-            threshold_z,
-        )
+class CompareRunsAPI(ApiEndpoint):
+    @staticmethod
+    def _get_all_results_for_a_run(run_id: str) -> List[BenchmarkResult]:
+        """Get all benchmark results for a run. Abort if the run doesn't exist or if
+        there are no results for the run.
+        """
+        result = Session.scalars(
+            s.select(BenchmarkResult).where(BenchmarkResult.run_id == run_id)
+        ).all()
 
-        return comparator.compare()
+        if not result:
+            f.abort(
+                404, description=f"no benchmark results found for run ID: '{run_id}'"
+            )
+
+        return list(result)
+
+    @staticmethod
+    def _join_results(
+        baseline_results: List[BenchmarkResult],
+        contender_results: List[BenchmarkResult],
+    ) -> List[Tuple[Optional[BenchmarkResult], Optional[BenchmarkResult]]]:
+        """Do a full outer join of two lists of benchmark results, pairing by case,
+        context, hardware, repo, and unit.
+
+        If a case/context/hardware/repo/unit combination is present in one list but not
+        the other, it will still be included but the other tuple element will be None.
+
+        If there are multiple results for a case/context/hardware/repo/unit combination
+        in both lists, a cartesian product of them all will be returned.
+        """
+        _key_type = Tuple[str, str, str, Optional[str], Optional[str]]
+
+        def _generate_key(result: BenchmarkResult) -> _key_type:
+            repo = result.run.commit.repository if result.run.commit else None
```

Comment:
IMO we should accept `"github": {"repo": <repo>}` instead of nothing for commitless results, but that's beyond the scope here probably

Reply:
Yeah, we will come back to this. #1165